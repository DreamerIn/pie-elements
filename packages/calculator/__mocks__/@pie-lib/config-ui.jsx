const config = {
  InputRadio: () => <div>Input Radio</div>,
  TwoChoice: () => <div>Two Choice</div>
}

module.exports = config;