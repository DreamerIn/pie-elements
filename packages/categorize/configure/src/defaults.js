export default {
  choices: [
    {
      id: '0',
      content: 'Choice 0'
    },
  ],
  choicesPerRow: 2,
  choicesPosition: 'below',
  choicesLabel: '',
  lockChoiceOrder: true,
  removeTilesAfterPlacing: false,
  categoriesPerRow: 2,

  categories: [
    {
      id: '0',
      label: 'Category 0',
      choices: []
    },
  ],
  correctResponse: [],
};
