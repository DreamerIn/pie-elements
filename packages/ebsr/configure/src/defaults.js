export default {
  partA: {
    choices: [
      {
        value: 'a',
        label: 'label a',
      }
    ],
    prompt: 'prompt a',
  },
  partB: {
    choices: [
      {
        value: 'b',
        label: 'label b',
      },
    ],
    prompt: 'prompt b',
  },
};
