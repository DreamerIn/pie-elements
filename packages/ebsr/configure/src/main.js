import { ModelUpdatedEvent } from '@pie-framework/pie-configure-events';
import MultipleChoiceConfigure from '@pie-element/multiple-choice/configure/lib';

import debug from 'debug';

const log = debug('pie-elements:ebsr:configure');

export default class Main extends HTMLElement {
  static defineMultipleChoice() {
    if(!customElements.get('multiple-choice-configure')){
      customElements.define('multiple-choice-configure', MultipleChoiceConfigure);
    }
  }

  static getParts = () => ({
    partA: document.getElementById('part-a-configure'),
    partB: document.getElementById('part-b-configure'),
  });

  constructor() {
    super();

    Main.defineMultipleChoice();
  }

  set model(m) {
    this._model = m;

    this._render();
    this.setupModel();
  }

  setupModel() {
    const { partA, partB } = Main.getParts();

    if (partA && partB) {
      this.setupPart(partA, 'partA');
      this.setupPart(partB, 'partB');
    }
  }

  setupPart(part, key) {
    this.initiatePart(part, key);
    this.captureModelUpdated(part, key);
  }

  initiatePart(part, key) {
    part.model = this._model[key];

    this.dispatchModelChanged();
  }

  dispatchModelChanged(reset = false) {
    log('[onModelChanged]: ', this._model);

    this.dispatchEvent(new ModelUpdatedEvent(this._model, reset));
  }

  captureModelUpdated(part, key) {
    const self = this;

    part.addEventListener('model.updated', event => {
      event.stopImmediatePropagation();
      self.repackModel(event.update, key);
    });
  }

  repackModel(partModel, key) {
    this._model =  {
      ...this._model,
      [key]: partModel
    };

    this.dispatchModelChanged();
  }

  _render() {
    this.innerHTML = `
      <div>
        <multiple-choice-configure id="part-a-configure"></multiple-choice-configure>
        <multiple-choice-configure id="part-b-configure"></multiple-choice-configure>
      </div>
    `;
  }
}
