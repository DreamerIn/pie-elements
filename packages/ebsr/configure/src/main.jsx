import React from 'react';
import PropTypes from 'prop-types';

import MultipleChoiceConfigure from '@pie-element/multiple-choice/configure/lib';

export default class Root extends React.Component {
  static propTypes = {
    model: PropTypes.object.isRequired,
    onModelChanged: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);

    if(!customElements.get('multiple-choice-configure')){
      customElements.define('multiple-choice-configure', MultipleChoiceConfigure);
    }

    this.state = {
      model: props.model,
    };
  }

  componentDidMount() {
    const { partA, partB } = this;

    customElements.whenDefined('multiple-choice-configure').then(() => {
      this.configurePart(partA, 'partA');
      this.configurePart(partB, 'partB');
    });
  }

  configurePart(part, key) {
    this.updatePart(part, key, true);
    this.handleModelUpdateListener(part, key);
  }

  updatePart(part, key, setModel) {
    const { onModelChanged } = this.props;
    const { model } = this.state;

    const { keyMode } = model;
    const { choices, choiceMode, labelMode, prompt } = model[key];

    const partModel = {
      prompt,
      choiceMode,
      keyMode,
      choices,
      labelMode,
      // hardcoded for now
      lockChoiceOrder: false,
      sequentialLabel: false
    };

    if (setModel) {
      part.model = partModel;
    }

    const repackedModel = {
      ...model,
      [key]: partModel,
    };

    onModelChanged(repackedModel);
  }

  handleModelUpdateListener(part, key) {
    const self = this;

    part.addEventListener('model.updated', (model) => {
      self.updatePart(model.update, key);
    })
  }

  render() {
    return (
      <div>
        <multiple-choice-configure ref={ref => { this.partA = ref }} />
        <multiple-choice-configure ref={ref => { this.partB = ref }} />
      </div>
    );
  }
}
