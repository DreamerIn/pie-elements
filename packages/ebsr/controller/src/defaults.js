export default {
  partA: {
    choices: [
      {
        value: 'a',
        label: 'label a',
      }
    ],
    prompt: 'prompt a',
    disabled: false,
  },
  partB: {
    choices: [
      {
        value: 'b',
        label: 'label b',
      },
    ],
    prompt: 'prompt b',
    disabled: false,
  },
};
