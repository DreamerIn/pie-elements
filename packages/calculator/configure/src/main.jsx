import React from 'react';
import ReactDOM from 'react-dom';
import Calculator from '@pie-framework/material-ui-calculator';
import { InputRadio, TwoChoice } from '@pie-lib/config-ui';

export default class Main extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      twoChoice: '',
      model: props.model
    };
  }

  render() {
    const { basic, scientific } = this.state;
    return (
      <div>
        <TwoChoice
          header="Choose Calculator Type"
          value={this.state.twoChoice}
          onChange={twoChoice => this.setState({ twoChoice })}
          one={{ label: 'Basic', value: 'basic' }}
          two={{ label: 'Scientific', value: 'scientific' }} />
          {this.state.twoChoice && <Calculator mode={this.state.twoChoice} />}
      </div>
    );
  }
}