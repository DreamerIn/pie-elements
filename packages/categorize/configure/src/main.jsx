import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import { Tabs } from '@pie-lib/config-ui';
import Design from './design';
import Scoring from './scoring';

export class Main extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    model: PropTypes.object.isRequired,
    onChange: PropTypes.func.isRequired,
    imageSupport: PropTypes.object
  };

  static defaultProps = {};

  changeScoring = scoring => {
    const { onChange, model } = this.props;
    model.scoring = scoring;
    onChange(model);
  };

  render() {
    const { classes, className, model, onChange } = this.props;

    if (!model.scoring) {
      model.scoring = {
        weighting: {
          enabled: false,
          rules: []
        },
        partial: {
          enabled: false,
          rules: []
        }
      };
    }

    return (
      <div className={classNames(classes.main, className)}>
        <Tabs>
          <Design
            imageSupport={this.props.imageSupport}
            title="Design"
            model={model}
            onChange={onChange}
          />
          <Scoring
            title="Scoring"
            scoring={model.scoring}
            categories={model.categories}
            correctResponse={model.correctResponse}
            onChange={this.changeScoring}
          />
        </Tabs>
      </div>
    );
  }
}
const styles = theme => ({
  main: {}
});

const StyledMain = withStyles(styles)(Main);

class Stateful extends React.Component {
  static propTypes = {
    onChange: PropTypes.func.isRequired,
    model: PropTypes.object.isRequired,
    imageSupport: PropTypes.object
  };

  constructor(props) {
    super(props);
    this.state = {
      model: props.model
    };
  }

  componentWillReceiveProps(props) {
    this.setState({ model: props.model });
  }

  onChange = model => {
    this.setState({ model }, () => {
      this.props.onChange(this.state.model);
    });
  };

  render() {
    return (
      <StyledMain
        imageSupport={this.props.imageSupport}
        model={this.state.model}
        onChange={this.onChange}
      />
    );
  }
}

export default Stateful;
