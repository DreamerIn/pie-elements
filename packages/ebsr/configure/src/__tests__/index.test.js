import React from 'react';
import { mount } from 'enzyme';
import { ModelUpdatedEvent } from '@pie-framework/pie-configure-events';
import { choiceUtils as utils } from '@pie-lib/config-ui';
import merge from 'lodash/merge';
import MultipleChoiceConfigure from '@pie-element/multiple-choice/configure/lib';
import ReactDOM from 'react-dom';
import defaults from '../defaults';
import { Main } from '../main';

jest.mock('react-dom', () => ({
  render: jest.fn()
}));

jest.mock('@pie-framework/pie-configure-events', () => ({
  ModelUpdatedEvent: (update) => {
    return {
      update,
      preventDefault: jest.fn(),
      stopImmediatePropagation: jest.fn()
    }
  }
}));

jest.mock('@pie-lib/config-ui', () => ({
  choiceUtils: {
    firstAvailableIndex: jest.fn()
  },
  settings: {
    Panel: props => <div {...props} />,
    toggle: jest.fn(),
    radio: jest.fn(),
    dropdown: jest.fn()
  },
  layout: {
    ConfigLayout: props => <div>{props.children}</div>
  }
}));

jest.mock(
  '@pie-element/multiple-choice/configure/lib',
  () => class MockConfigure {}
);

const PART_A = 'partA';
const PART_B = 'partB';

const model = {
  partA: {
    choiceMode: 'radio',
    choices: [
      {
        value: 'yellow',
        label: 'Yellow',
        correct: true,
        feedback: {
          type: 'custom',
          value: 'foo'
        }
      },
      {
        value: 'green',
        label: 'Green',
        feedback: {
          type: 'default'
        }
      }
    ],
    choicePrefix: 'numbers',
    prompt: `prompt ${PART_A}`
  },
  partB: {
    choiceMode: 'radio',
    choices: [
      {
        value: 'orange',
        label: 'Orange',
        correct: true,
        feedback: {
          type: 'custom',
          value: 'foo'
        }
      },
      {
        value: 'purple',
        label: 'Purple',
        feedback: {
          type: 'default'
        }
      }
    ],
    choicePrefix: 'numbers',
    prompt: `prompt ${PART_B}`
  }
};

const createDefaultModel = (model = {}) => ({
  ...defaults.model,
  ...model,
  partA: {
    ...defaults.model.partA,
    ...model.partA
  },
  partB: {
    ...defaults.model.partB,
    ...model.partB
  },
});

describe('index', () => {
  let Def;
  let el;
  let onModelChanged = jest.fn();
  let onConfigurationChanged = jest.fn();
  let main;

  beforeAll(() => {
    Def = require('../index').default;
  });

  beforeEach(() => {
    el = new Def();
    el.model = model;
    el.onModelChanged = onModelChanged;
    el.onConfigurationChanged = onConfigurationChanged;
    el.connectedCallback();
    el.dispatchEvent = el.onModelUpdated;

    main = mount(<Main
      classes={{}}
      model={el._model}
      configuration={defaults.configuration}
      onConfigurationChanged={onConfigurationChanged}
      onModelChanged={onModelChanged}
    />);

    // mock onModelChanged to dispatch a MODEL_UPDATED event (as multiple-choice does)
    main.instance().partA.onModelChanged = (m) => {
      const event = new ModelUpdatedEvent(m, false);

      el.dispatchEvent({
        ...event,
        target: {
          getAttribute: jest.fn().mockReturnValue('A')
        }
      });
    };
    main.instance().partB.onModelChanged = (m) => {
      const event = new ModelUpdatedEvent(m, false);

      el.dispatchEvent({
        ...event,
        target: {
          getAttribute: jest.fn().mockReturnValue('B')
        }
      });
    };
  });

  describe('set model', () => {
    it('calls ReactDOM.render', () => {
      el.model = model;

      expect(ReactDOM.render).toHaveBeenCalled();
      expect(el._model).toEqual(createDefaultModel(model));
    });
  });

  describe('set configuration', () => {
    it('calls ReactDOM.render', () => {
      el.configuration = defaults.configuration;

      expect(ReactDOM.render).toHaveBeenCalled();
    });
  });

  const assetOnModelUpdated = (label, key, updatedPart, expected) => {
    it(label, () => {
      const event = new ModelUpdatedEvent(updatedPart, false);

      el.dispatchEvent({
        ...event,
        target: {
          getAttribute: jest.fn().mockReturnValue(key)
        }
      });

      expect(ReactDOM.render).toBeCalled();
      expect(el._model[`part${key}`]).toEqual(expected);
    });
  };

  describe('onModelUpdated', () => {
    assetOnModelUpdated(
      'dispatching MODEL_UPDATED updates model.partA',
      'A',
      { updatedA: true },
      { updatedA: true }
      );
    assetOnModelUpdated(
      'dispatching MODEL_UPDATED updates model.partB',
      'B',
      { updatedB: true },
      { updatedB: true });

    assetOnModelUpdated(
      'dispatching MODEL_UPDATED with update undefined does not update model.partA',
      'A',
      undefined,
      createDefaultModel(model).partA);
    assetOnModelUpdated(
      'dispatching MODEL_UPDATED with update undefined does not update model.partB',
      'B',
      undefined,
      createDefaultModel(model).partB);
  });




  const assetOnModelChanged = (label, key, updatedModel) => {
    it(label, () => {
      main.instance().props.onModelChanged({ test: 'a' });

      expect(onModelChanged).toHaveBeenCalledWith({ test: 'a' });
    });
  };


  const assetPartUpdate = (label, key, updatedModel) => {
    it(label, () => {
      main.instance().partA.onModelChanged({ test: 'ceva'});

      console.log('assetPartUpdate');
      expect(ReactDOM.render).toBeCalled();
    });
  };
});


// these tests are no longer proper for EBSR, since previously React was not used
xdescribe('index', () => {
  let Def;
  let el;
  let ebsr;

  beforeAll(() => {
    Def = require('../index').default;
  });

  beforeEach(() => {
    el = new Def();
    el.connectedCallback();
    el.populatePart = jest.fn();
    ebsr = {
      partA: new HTMLElement(),
      partB: new HTMLElement()
    };
    el.querySelector = jest.fn(s => {
      if (s === `ebsr-multiple-choice-configure#a`) {
        return ebsr.partA;
      } else {
        return ebsr.partB;
      }
    });
    el.model = model;
  });

  const shouldHaveModel = key => {
    it(`${key} should have set the model`, () => {
      expect(ebsr[key].model).toEqual(model[key]);
    });
  };

  const updateModel = (model, key) => {
    const event = new ModelUpdatedEvent(model, false);

    el._model[key] = event.update;
    el.dispatchEvent(new ModelUpdatedEvent(el._model, false));
  };

  const expectToMatchModel = (newModel, key) => {
    expect(newModel).toEqual(model[key]);
  };

  const resetsModel = key => {
    it(`${key} resets the model`, () => {
      const newModel = {
        ...model[key],
        choiceMode: 'checkbox'
      };

      updateModel(newModel, key);
      expectToMatchModel(newModel, key);
    });
  };

  const changesPartialScoring = key => {
    it(`${key} changes partial scoring value`, () => {
      const newModel = {
        ...model[key],
        partialScoring: true
      };

      updateModel(newModel, key);
      expectToMatchModel(newModel, key);
    });
  };

  const addsChoice = key => {
    it(`${key} adds a choice`, () => {
      const newModel = {
        ...model[key],
        choices: [
          ...model[key].choices,
          {
            label: 'label',
            value: utils.firstAvailableIndex(
              model[key].choices.map(c => c.value),
              0
            ),
            feedback: {
              type: 'none'
            }
          }
        ]
      };

      updateModel(newModel, key);
      expectToMatchModel(newModel, key);
    });
  };

  const removesChoice = key => {
    it(`${key} removes choice`, () => {
      const newModel = {
        ...model[key],
        choices: model[key].choices.splice(0, 2)
      };

      updateModel(newModel, key);
      expectToMatchModel(newModel, key);
    });
  };

  const changesChoicePrefix = key => {
    it(`${key} changes choicePrefix`, () => {
      const newModel = {
        ...model[key],
        choicePrefix: 'letters'
      };

      updateModel(newModel, key);
      expectToMatchModel(newModel, key);
    });
  };

  const changesChoice = (key, value) => {
    const choice = {
      correct: true,
      value,
      label: value => value.charAt(0).toUpperCase() + value.slice(1),
      feedback: {
        type: 'none',
        value: ''
      }
    };

    it(`${key} changes choice`, () => {
      const newModel = {
        ...model[key],
        choiceMode: 'checkbox'
      };

      newModel.choices.splice(1, 1, choice);

      updateModel(newModel, key);
      expectToMatchModel(newModel, key);
    });

    it(`${key} changes choice and makes incorrect all other choices`, () => {
      const newModel = {
        ...model[key],
        choiceMode: 'radio',
        choices: [
          ...model[key].choices.map(c => {
            return merge({}, c, { correct: false });
          })
        ]
      };

      newModel.choices.splice(1, 1, choice);

      updateModel(newModel, key);
      expectToMatchModel(newModel, key);
    });
  };

  shouldHaveModel(PART_A);
  shouldHaveModel(PART_B);

  describe('logic', () => {
    describe('onChoiceModeChanged', () => {
      resetsModel(PART_A);
      resetsModel(PART_B);
    });

    describe('onPartialScoringChanged', () => {
      changesPartialScoring(PART_A);
      changesPartialScoring(PART_B);
    });

    describe('onAddChoice', () => {
      addsChoice(PART_A);
      addsChoice(PART_B);
    });

    describe('onRemoveChoice', () => {
      removesChoice(PART_A);
      removesChoice(PART_B);
    });

    describe('onKeyModeChanged', () => {
      changesChoicePrefix(PART_A);
      changesChoicePrefix(PART_B);
    });

    describe('onChoiceChanged', () => {
      changesChoice(PART_A, 'green');
      changesChoice(PART_B, 'purple');
    });
  });
});
