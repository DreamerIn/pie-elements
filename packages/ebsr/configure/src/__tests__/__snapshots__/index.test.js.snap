// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EBSR Configure createDefaultModel default-snapshot 1`] = `
Object {
  "partA": Object {
    "choiceMode": "radio",
    "choicePrefix": "numbers",
    "choices": Array [],
    "feedbackEnabled": true,
    "partialScoring": false,
    "prompt": "Prompt A",
    "promptEnabled": true,
    "rationale": "",
    "rationaleEnabled": true,
    "studentInstructionsEnabled": true,
    "teacherInstructions": "",
    "teacherInstructionsEnabled": true,
  },
  "partB": Object {
    "choiceMode": "radio",
    "choicePrefix": "numbers",
    "choices": Array [],
    "feedbackEnabled": true,
    "partialScoring": false,
    "prompt": "Prompt B",
    "promptEnabled": true,
    "rationale": "",
    "rationaleEnabled": true,
    "studentInstructionsEnabled": true,
    "teacherInstructions": "",
    "teacherInstructionsEnabled": true,
  },
  "partLabelType": "Letters",
  "partLabels": true,
}
`;

exports[`EBSR Configure createDefaultModel with-overrides-snapshot 1`] = `
Object {
  "partA": Object {
    "choiceMode": "radio",
    "choicePrefix": "numbers",
    "choices": Array [],
    "feedbackEnabled": true,
    "partialScoring": false,
    "prompt": "Prompt A",
    "promptEnabled": true,
    "rationale": "foo",
    "rationaleEnabled": true,
    "studentInstructionsEnabled": true,
    "teacherInstructions": "ti",
    "teacherInstructionsEnabled": true,
  },
  "partB": Object {
    "choiceMode": "radio",
    "choicePrefix": "numbers",
    "choices": Array [],
    "feedbackEnabled": true,
    "partialScoring": false,
    "prompt": "Prompt B",
    "promptEnabled": true,
    "rationale": "",
    "rationaleEnabled": true,
    "studentInstructionsEnabled": true,
    "teacherInstructions": "",
    "teacherInstructionsEnabled": true,
  },
  "partLabelType": "Letters",
  "partLabels": true,
}
`;
